optimization/src/node_local.rs

```diff
@@ -0,0 +1,159 @@
+//! Some simple node local optimizations, e.g.
+//! - algebraic simplifications,
+//! - div of power of two to shift
+use super::Outcome;
+use crate::optimization;
+use libfirm_rs::{
+    nodes::{Node, NodeTrait},
+    Graph, Mode, Tarval, TarvalKind,
+};
+
+pub struct NodeLocal {
+    graph: Graph,
+    changed: Outcome,
+}
+
+impl optimization::Local for NodeLocal {
+    fn optimize_function(graph: Graph) -> Outcome {
+        NodeLocal::new(graph).run()
+    }
+}
+
+impl NodeLocal {
+    fn new(graph: Graph) -> Self {
+        Self {
+            graph,
+            changed: Outcome::Unchanged,
+        }
+    }
+
+    fn visit_node(&mut self, current_node: Node) {
+        // do not put a breakpoint here, this will result in a walker in walker
+        // which just does random shit in libfirm.
+        log::debug!("Div2Shift: visiting {:?}", current_node);
+
+        #[allow(clippy::single_match)]
+        match current_node {
+            Node::Div(div) => {
+                log::debug!(
+                    "Div2Shift: {:?}[left:{:?},right:{:?}] ",
+                    div,
+                    div.left(),
+                    div.right(),
+                );
+                if let Node::Conv(conv) = div.right() {
+                    log::debug!("Div2Shift: conv {:?}!", conv);
+                    if let Node::Const(divisor) = conv.op() {
+                        log::debug!("Div2Shift: divisor {:?}!", divisor);
+                        if let TarvalKind::Long(divisor_value) = divisor.tarval().kind() {
+                            log::debug!("Div2Shift: divisor value {:?}!", divisor_value);
+                            if is_abs_power_of_two(divisor_value) {
+                                let has_minus = divisor_value < 0;
+
+                                log::debug!("Div2Shift: is power of two!");
+                                let shift_amount = (divisor_value.abs() as f64).log2() as i64;
```

Comment:
Can this ever not work due to floating point precision? You should be able to just do an abs, then get the highest bit position